wvd-templates/DSC/Script-RegisterSessionHost.ps1

```diff
@@ -93,39 +93,55 @@ else {
-[Microsoft.RDInfra.RDManagementData.RdMgmtSessionHost]$rdsh = $pssh.GetSessionHost()
-Write-Log -Message "RDSH object content: `n$($rdsh | Out-String)"
+Write-Log -Message "Getting RDSH session host info for '$SessionHostName'"
+
+# Wait for the session host to be available
+$StartTime = Get-Date
+$SessionHost = $null
+$TimeOutInSec = 900
+$DesiredStates = ('Available', 'NeedsAssistance')
+
+Write-Log -Message "Wait for sessions host to be in any of the desired states: $($DesiredStates -join ', ')"
+while ((!$SessionHost -or $SessionHost.Status -notin $DesiredStates) -and (get-date).Subtract($StartTime).TotalSeconds -lt $TimeOutInSec) {
+    write-log "Session host is in '$($SessionHost.Status)' state, continue waiting"
+    Start-Sleep -Seconds 30
+    $SessionHost = Get-RdsSessionHost -TenantName $TenantName -HostPoolName $HostPoolName -Name $SessionHostName -ErrorAction SilentlyContinue
+}
 
-$rdshName = $rdsh.SessionHostName | Out-String -Stream
-$poolName = $rdsh.hostpoolname | Out-String -Stream
+Write-Log -Message "RDSH object content: `n$($SessionHost | Out-String)"
 
-Write-Log -Message "Waiting for session host return when in available status"
-$AvailableSh = $pssh.GetSessionHostWhenAvailable()
-if ($null -ne $AvailableSh) {
-    Write-Log -Message "Session host $($rdsh.SessionHostName) is now in Available state"
+if (!$SessionHost) {
+    throw "RD Agent failed to register VM '$SessionHostName' to HostPool '$HostPoolName'"
+}
+if ($SessionHost.Status -in $DesiredStates) {
+    Write-Log -Message "Session host is now in '$($SessionHost.Status)' state"
 }
 else {
-    Write-Log -Err "Session host $($rdsh.SessionHostName) not in Available state, wait timed out (threshold is $($rdsh.TimeoutInSec) seconds)"
+    Write-Log -Err "Session host is not in any of the desired states: $($DesiredStates -join ', '). Wait timed out (threshold is $TimeOutInSec seconds)"
 }
 
-# check if the session host was successfully registered to host pool, note that the error is thrown because the TestScript configuration of DSC will not be run after SetScript (this script)
-$IsSessionHostRegisterd = (& "$ScriptPath\Script-TestRegisterSessionHost.ps1" -RdBrokerURL $RDBrokerURL -DefinedTenantGroupName $definedTenantGroupName -TenantName $TenantName -HostPoolName $HostPoolName -TenantAdminCredentials $TenantAdminCredentials -isServicePrincipal $isServicePrincipal -aadTenantId $AadTenantId -RDPSModSource $RDPSModSource)
-if (!$IsSessionHostRegisterd) {
-    throw "RD Agent failed to register $rdshName VM to $poolName"
+# check if the session host was successfully registered to host pool, note that the error is thrown because the TestScript configuration of DSC may not be run after SetScript (this script)
+Write-Log -Message "Check RD Infra registry to see if RD Agent is registered"
+$RDInfraReg = Get-ItemProperty -Path "Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\RDInfraAgent" -ErrorAction SilentlyContinue

```

Comment:
These registry checks seem to be the same as the ones in Script-TestRegisterSessionHost.ps1 except that it throws instead of returning false. Can it be centralized to a function?

Reply:
good idea